src/protocol/iobuf.c

```diff
@@ -439,27 +436,48 @@ int _z_wbuf_write_bytes(_z_wbuf_t *wbf, const uint8_t *bs, size_t offset, size_t
         _z_iosli_write_bytes(ios, bs, offset, length);
         return 0;
     }
-    else if (wbf->is_expandable)
+
+    if (wbf->is_expandable)
     {
-        // Write what left
-        if (writable > 0)
+        _z_iosli_write_bytes(ios, bs, offset, writable);
+        length -= writable;
+        offset += writable;
+        while (length > 0)
         {
+            __z_wbuf_new_iosli(wbf, ZN_IOSLICE_SIZE);
+
+            wbf->w_idx++;
+            ios = _z_wbuf_get_iosli(wbf, wbf->w_idx);
+
+            writable = _z_iosli_writable(ios);
+            if (length < writable)
+                writable = length;
+
             _z_iosli_write_bytes(ios, bs, offset, writable);
-            offset += writable;
             length -= writable;
+            offset += writable;
         }
 
-        // Allocate N capacity slots to hold at least length bytes
-        size_t capacity = wbf->capacity * (1 + (length / wbf->capacity));
-        __z_wbuf_new_iosli(wbf, capacity);
-        ios = _z_wbuf_get_iosli(wbf, wbf->w_idx);
-        _z_iosli_write_bytes(ios, bs, offset, length);
         return 0;
     }
-    else
-    {
-        return -1;
-    }
+
+    return -1;
+}
+
+int _z_wbuf_wrap_bytes(_z_wbuf_t *wbf, const uint8_t *bs, size_t offset, size_t length)
+{
+    _z_iosli_t *ios = _z_wbuf_get_iosli(wbf, wbf->w_idx);
+    size_t writable = _z_iosli_writable(ios);
+    ios->capacity = ios->w_pos; // Block ioslice
+
+    wbf->w_idx++;
+    _z_iosli_t wios = _z_iosli_wrap(bs + offset, length, 0, length);
+    _z_wbuf_add_iosli(wbf, _z_iosli_clone(&wios));
+
+    wbf->w_idx++;
```

Comment:
`w_idx` increment should be handled by the proper `wbuf` function adding the `iosli`.